### PR TITLE
Fixed #32552 -- Added logger argument to DiscoverRunner.

### PR DESCRIPTION
**Pull request by cjerdonek**
https://code.djangoproject.com/ticket/32552

### CODE REVIEW

File: django/test/runner.py [hunk 1]
```diff
@@ -671,16 +672,23 @@ def shuffle_seed(self):
 
     def log(self, msg, level=None):
         """
-        Log the given message at the given logging level.
+        Log the message at the given logging level (default of INFO).
 
-        A verbosity of 1 logs INFO (the default level) or above, and verbosity
-        2 or higher logs all levels.
+        If a logger isn't set, the message is instead printed to the console
+        according to the configured verbosity. A verbosity of 1 prints INFO and
```
Comment (smithdc1) [inline django/test/runner.py:678]: Nit pick. Maybe order these 0, 1, 2+.

File: docs/topics/testing/advanced.txt [hunk 2]
```diff
@@ -580,10 +580,15 @@ and tear down the test suite.
...
-    using this option.
+    be logged and set to ``self.shuffle_seed`` prior to running tests. This
+    option can be used to help detect tests that aren't properly isolated.
+    :ref:`Grouping by test class <order-of-tests>` is preserved when using this
+    option.
+
+    ``logger`` can be used to pass a Python `Logger object`_. If provided, the
+    Logger will be used to log messages instead of printing to the console. The
+    Logger object will respect its logging level rather than the configured
+    ``verbosity``.
```
Comment (smithdc1) [inline docs/topics/testing/advanced.txt:591]: Missing a `..versionadded`?

I also wonder if a link to the refreshed logging docs could be helpful?

https://github.com/django/django/blob/main/docs/howto/logging.txt

File: docs/topics/testing/advanced.txt [hunk 3]
```diff
@@ -580,10 +580,15 @@ and tear down the test suite.
...
-    be logged to the console and set to ``self.shuffle_seed`` prior to running
-    tests. This option can be used to help detect tests that aren't properly
-    isolated. :ref:`Grouping by test class <order-of-tests>` is preserved when
-    using this option.
+    be logged and set to ``self.shuffle_seed`` prior to running tests. This
+    option can be used to help detect tests that aren't properly isolated.
+    :ref:`Grouping by test class <order-of-tests>` is preserved when using this
+    option.
+
+    ``logger`` can be used to pass a Python `Logger object`_. If provided, the
```
Comment (smithdc1) [inline docs/topics/testing/advanced.txt:588]: Can use intersphinx to link to Python docs? 

E.g. in the logging docs we use :py:func:`logging.getLogger`.

File: docs/topics/testing/advanced.txt (hunk 2 above)
Comment (cjerdonek) [reply docs/topics/testing/advanced.txt]: Thanks for your comments, @smithdc1. I incorporated all of them except for linking to the how-to docs. I'm not sure it's needed for this niche argument, and I didn't see a way to add a link without adding verbiage like, "For information on configuring and using logging, see ...," which seemed like clutter here. If you have something in mind that's unobtrusive, feel free to suggest.

Comment (smithdc1) [reply docs/topics/testing/advanced.txt]: That's fine, it was just a thought. 👍

File: docs/topics/testing/advanced.txt [hunk 4]
```diff
@@ -601,7 +606,9 @@ and tear down the test suite.
 
     .. versionadded:: 4.0
 
-        The ``shuffle`` argument was added.
+        The ``logger`` and ``shuffle`` arguments were added.
+
+.. _`Logger object`: https://docs.python.org/3/library/logging.html#logger-objects
```
Comment (smithdc1) [inline docs/topics/testing/advanced.txt:611]: It's this link still required?

File: docs/topics/testing/advanced.txt [hunk 5]
```diff
@@ -726,11 +731,13 @@ Methods
...
-    respecting the current ``verbosity``. For example, an ``INFO`` message will
-    be logged if the ``verbosity`` is at least 1, and ``DEBUG`` will be logged
-    if it is at least 2.
+    If a ``logger`` is set, logs the message at the given integer
+    `logging level`_ (e.g. ``logging.DEBUG``, ``logging.INFO``, or
+    ``logging.WARNING``). Otherwise, the message is printed to the console,
+    respecting the current ``verbosity``. For example, no message will be
+    printed if the ``verbosity`` is 0, ``INFO`` and above will be printed if
+    if the ``verbosity`` is at least 1, and ``DEBUG`` will be printed if
+    it is at least 2. The ``level`` defaults to ``logging.INFO``.
```
Comment (felixxm) [inline docs/topics/testing/advanced.txt]:
```suggestion
    printed if the ``verbosity`` is 0, ``INFO`` and above will be printed if
    the ``verbosity`` is at least 1, and ``DEBUG`` will be printed if it is at
    least 2. The ``level`` defaults to ``logging.INFO``.
```

**Review by felixxm — COMMENTED** [review]
@cjerdonek Thanks :+1: I pushed minor edits to docs.

File: docs/topics/testing/advanced.txt (hunk 5 above)
Comment (cjerdonek) [reply docs/topics/testing/advanced.txt]: Thanks!

**Pull request merged**